kart/annotations/cli.py

```diff
@@ -0,0 +1,74 @@
+import click
+import pygit2
+
+from .db import annotations_session
+from kart.diff_estimation import estimate_diff_feature_counts
+
+EMPTY_TREE_SHA = "4b825dc642cb6eb9a060e54bf8d69288fbee4904"
+
+
+def gen_reachable_commits(repo):
+    """
+    Generator. Yields commits that are reachable from any ref.
+    """
+    refs = repo.references
+    commit_queue = set()
+    commit_shas_seen = set()
+    for k in refs:
+        commit_queue.add(refs[k].peel(pygit2.Commit))
+
+    while commit_queue:
+        commit = commit_queue.pop()
+        if commit.oid.hex in commit_shas_seen:
+            continue
+        try:
+            parents = commit.parents
+        except KeyError:
+            # shallow repo. no big deal
+            parents = []
+        for p in parents:
+            commit_queue.add(p)
+        commit_shas_seen.add(commit.oid.hex)
+        yield commit
+
+
+@click.command(name="build-annotations")
+@click.pass_context
+@click.option(
+    "--all-reachable", is_flag=True, help="Build annotations for all reachable commits"
+)
+@click.argument(
```

Comment:
Presumably a common way to use this is in a post-receive hook. Normally it's the hook's responsibility to resolve pushed refs to a set of commit IDs which can then be passed to `build-annotations`, but there could be a a lot. Maybe a way to accept refishes via stdin?

Reply:
This is probably _only_ useful as a low-level command to be run in hooks; users aren't going to run it. I'll change it so the only way to send it IDs is via stdin, or use `--all-reachable`

Reply:
seems sane 👍